src/tetris/biasEngine/biasEvent.ts

```diff
@@ -16,25 +31,121 @@ export default class BiasEvent {
 	public get endTime(): number {
 		return this._endTime;
 	}
+
+	public get totalDetectionLevelIncrease(): number {
+		return ((this.endTime - this.startTime) / 1000.0) * this._detectionLevelIncreasePerSecond;
+	}
+
+	public get isActive(): boolean {
+		const now = Date.now();
+
+		if (now < this.startTime) {
+			return false;
+		}
+		else if (now > this.endTime) {
+			return false;
+		}
+
+		return true;
+	}
 	//endregion
 
 	//region public methods
+	public initializeFromPrototype(detectionLevelDelta: number): BiasEvent {
+		let clone = this._clone();
+		const duration = this.durationRange * (detectionLevelDelta / this._detectionLevelIncreasePerSecond);
+
+		const now = Date.now();
+		clone._startTime = now;
+		clone._endTime = now + Utility.limitValueBetweenMinAndMax(duration, this.minDurationInMs, this.maxDurationInMs);
+
+		return clone;
+	}
+
+	public calculateSpawnProbability(currentBias: number) {
+		const relativeBias = currentBias - BiasEngine.NEUTRAL_BIAS_VALUE;
+		const relativeThreshold = this._spawnBiasThreshold - BiasEngine.NEUTRAL_BIAS_VALUE;
+
+		if (relativeBias == 0) {
+			return 0.0;
+		}
+
+		if (!this._haveSameSign(relativeBias, relativeThreshold)) {
+			return 0.0;
+		}
+
+		if (Math.abs(relativeBias) >= Math.abs(relativeThreshold)) {
+			return 1.0;
+		}
+
+		return Math.abs(relativeThreshold) / Math.abs(relativeBias);
+	}
+
+	public calculateDetectionLevelDecrease(deltaInMs: number): number {
+		return (deltaInMs / 1000.0) * this._detectionLevelDecreasePerSecond;
+	}
 	//endregion
 
 	//region constructor
-	public constructor(biasEventType: BiasEventType, startTime: number, endTime: number) {
+	protected constructor(biasEventType: BiasEventType) {
 		this._eventType = biasEventType;
-		this._startTime = startTime;
-		this._endTime = endTime;
 	}
 	//endregion
 
 	//region private members
 	private readonly _eventType: BiasEventType;
-	private readonly _startTime: number;
-	private readonly _endTime: number;
+	private _startTime: number;
+	private _endTime: number;
 	//endregion
 
 	//region private methods
+	private _haveSameSign(value1: number, value2: number): boolean {
```

Comment:
That function should get outsourced to Utils.

`return value1 * value2 => 0` would be way shorter